stdlib/universe/join.go

```diff
@@ -697,7 +699,25 @@ func (c *MergeJoinCache) insertIntoBuffer(id execute.DatasetID, tbl flux.Table)
 
 		c.intersection = intersection
 	}
+
+	// Optimization: if any group key columns overlap join key columns,
+	// and there are any nulls in those columns, we can discard this table,
+	// since null != null for joining purposes.
+	k := tbl.Key()
+	for j, col := range k.Cols() {
+		if c.on[col.Label] {
+			if k.IsNull(j) {
+				// Discard the table and return.
+				err := tbl.Do(func(flux.ColReader) error {
```

Comment:
Why is the explicit call to `tbl.Do` necessary? Can we not just return right here, not caching the table?

Reply:
Apparently we need to iterate over the table for it to be consumed.  Lorenzo filed an issue for this:
https://github.com/influxdata/flux/issues/643
I'm not quite clear on why this is exactly, though.  There are a few different implementions of `flux.Table`, I think some of them may require their contents to be iterated exactly once for the pipeline to work right.

I'll add a comment to the code.